ui/radar/radar.js

```diff
@@ -0,0 +1,251 @@
+'use strict';
+
+let Options = {
+  Language: 'en',
+  UpdateMonsterPos: false,
+  DetectionRange: 0,
+  OnlyMobs: true,
+  TTS: false,
+  PopSoundAlert: true,
+  PopSound: '../../resources/sounds/PowerAuras/sonar.ogg',
+  PopVolume: 0.5,
+  Puller: false,
+  Position: false,
+};
+
+let gRadar;
+
+let instanceChangedRegex = {
+  'en': /00:0039:You are now in the instanced area/,
+  'cn': /00:0039:当前所在副本区为/,
+  'de': /00:0039:Du bist nun in dem instanziierten Areal/,
+  'fr': /00:0039:Vous avez été transporté/,
+  'ja': /00:0039:インスタンスエリア/,
+};
+
+class Point2D {
+  constructor(x, y) {
+    this.x = x;
+    this.y = y;
+  }
+}
+
+function length(a) {
+  return Math.sqrt((a.x) * (a.x) + (a.y) * (a.y));
+}
+
+function dot(a, b) {
+  return a.x * b.x + a.y * b.y;
+}
+
+function theta(a, b) {
+  return dot(a, b)/(length(a) * length(b));
+}
+
+function posToMap(h) {
+  let Offset = 21.5;
+  let Pitch = 0.02;
+  return (h * Pitch) + Offset;
+}
+
+class Radar {
+  constructor(element) {
+    this.targetMonsters = {};
+    this.playerPos = {};
+    this.table = element;
+    this.options = Options;
+    this.monsters = Object.assign({}, gMonster, Options.CustomMonsters);
+    this.lang = this.options.Language || 'en';
+    this.nameToMonster = {};
+    for (let i in this.monsters) {
+      let monster = this.monsters[i];
+      this.nameToMonster[monster['name'][this.lang] || monster['name']['en']] = monster;
+    }
+  }
+
+  AddMonster(log, monster, matches) {
+    let lang = this.lang;
+    let matchOrNot = (matches.groups.name.match(monster['name'][lang] || monster['name']['en']) != null);
+    matchOrNot &= (log.match(monster['regex'] || '') != null);
+    matchOrNot &= (parseFloat(matches.groups.hp) >= (monster['hp'] || 0));
+    let options = this.options;
+    if (monster['rank'] in options.RankOptions) // options overwrite
+      options = Object.assign({}, this.options, options.RankOptions[monster['rank']]);
+    if (options.OnlyMobs) {
+      matchOrNot &= matches.groups.id.startsWith('4');
+      matchOrNot &= !(log.match(/\(\d{0,5}00000\d{0,5}\)\.$/) === null);
```

Comment:
I added an `npcId` here for this to the addedCombatant regex so you don't have to match it manually.

Reply:
cool!

Reply:
got it